### ISSUE TITLE
Different results using pyro-ppl 1.5.1 vs pyro-ppl 1.6. batch & event shapes

### ISSUE THREAD

**Issue by LysSanzMoreta** [pyro-ppl/pyro#2787]
Hi!!!!

Pyro : 1.6
torch : 1.8.0                   
torchaudio : 0.7.0   
torchvision : 0.8.1            

I have a model that was doing alright, for a misterious reason I decided to upgrade to Pyro-ppl 1.6. This "smart" move forced me to introduce a plate in my model, otherwise it would not run. This simple change made my model underperform. I am trying to understand what was the default behaviour in previous pyro versions and how to reproduce it with a plate (or else) in pyro 1.6. I am using always the same seed. There is only 1 batch, I am not splitting my dataset (it's very small, no needed for now)

I have read many times : https://pyro.ai/examples/tensor_shapes.html and do not fully understand the differences in plates that apparently look the same.

```
Inference method: SVI
Guide: AutoDiagonalNormal
#n_seq = Number of sequences
#max_len = Largest sequence
#Prob = probabilities
```

Previous working model in Pyro 1.5.1

```py
def model_Old():
     ....
     logits = decoder(...) #[n_seq,max_len,prob]
     pyro.sample("sequences", dist.Categorical(logits=logits), obs=sequences) #[n_seq,max_len]
```

If I have understood correctly, here the "sequences" distribution shape does not have any dependent dimension, all are in the batch_shape. 

New models in Pyro 1.6

```py
def modelA():
     ....
     with pyro.plate("plate_sequences", size = max_len, dim=-3):
           logits = decoder(...) #[n_seq,max_len,prob]
           pyro.sample("sequences", dist.Categorical(logits=logits), obs=sequences) #[n_seq,max_len]
```

```py
def modelB():
     ....
    with pyro.plate("plate_sequences", size=n_seq, dim=-3):
           logits = decoder(...) #[n_seq,max_len,prob]
           pyro.sample("sequences", dist.Categorical(logits=logits), obs=sequences) #[n_seq,max_len]
```
```py
def modelC():
     ....
     with pyro.plate("plate_sequences",size=n_seq, dim=-2):
           logits = decoder(...) #[n_seq,max_len,prob]
           pyro.sample("sequences", dist.Categorical(logits=logits), obs=sequences) #[n_seq,max_len]
```

I am not sure on the exact differences between those plates (modelA vs modelB vs modelC). They have the same event_shape = [] and batch_shape = [n_seq, max_len] but they yield different results and none of them are as good as back in Pyro 1.5.1. 

Thanks in advance for any input,

Best wishes :)

**Comment by fritzo** [comment]
Hi @LysSanzMoreta, one recent change is that Pyro and PyTorch now **enable validation by default**.  Whereas it would previously have worked in some models to avoid both plates and `.to_event()`, Pyro will now give an error.  To temporarily revert to the old behavior you could try to `pyro.enable_validation(False)`.  To work with the new behavior, you could either add plates or add `.to_event(...)` in various places in your code.  Since adding plates appears to hurt performance I'd recommend trying to add `.to_event(...)` to your old model, e.g.:
```py
def modelD():
    ....
    logits = decoder(...) #[n_seq,max_len,prob]
    pyro.sample("sequences",
                dist.Categorical(logits=logits).to_event(sequences.dim()),
                obs=sequences) #[n_seq,max_len]
```
If you do want to use plates, I think you may need two plates, but let me know if I've misunderstood:
```py
def modelE():
    ....
    logits = decoder(...) #[n_seq,max_len,prob]
    with pyro.plate("seq", n_seq, dim=-2), pyro.plate("len", max_len, dim=-1):
        pyro.sample("sequences", dist.Categorical(logits=logits),
                    obs=sequences) #[n_seq,max_len]
```

**Comment by LysSanzMoreta** [comment]
@fritzo  Thanks for the quick reply :). I did not notice that `pyro.enable_validation `was activated by default, ops. I will run it with validation inactivated and the old model and hopefully it should be ok. Otherwise I will look into 2 plates. And I will check again `to_event() `, but I think it did not help either. Thanks (cause I am desperate hahah) !!!
